### PR TITLE
Initial work on {Create,Describe,Rename,Delete,List}Table(s).

### PR DESCRIPTION
**Pull request by petermattis**

### CODE REVIEW

File: client/namespace.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+		descKey := keys.MakeDescMetadataKey(desc.ID)
+		b := &Batch{}
+		b.CPut(nameKey, descKey, nil)
+		b.Put(descKey, &desc)
+		return txn.Commit(b)
+	})
+}
+
+// DescribeTable retrieves the table schema for the specified table.
+func (db *DB) DescribeTable(name string) (proto.TableSchema, error) {
```
Comment (tamird) [inline client/namespace.go]: maybe this should return a pointer so that it can be nil in the error cases?

File: client/namespace.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+	if !gr.Exists() {
+		return fmt.Errorf("unable to find table \"%s\"", name)
+	}
+	descKey := gr.ValueBytes()
+	desc := proto.TableDescriptor{}
+	if err := db.GetProto(descKey, &desc); err != nil {
+		return err
+	}
+
+	panic("TODO(pmattis): delete all of the tables rows")
```
Comment (tamird) [inline client/namespace.go]: nice

File: client/namespace.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+	// return db.Del(descKey)
+}
+
+// ListTables lists the tables.
+func (db *DB) ListTables() ([]string, error) {
+	tableNamePrefix := keys.MakeNameMetadataKey(0, "")
```
Comment (tamird) [inline client/namespace.go]: this should probably be a constant

File: keys/keys.go [hunk 4]
```diff
@@ -56,17 +56,20 @@ func NodeStatusKey(nodeID int32) proto.Key {
 	return MakeKey(StatusNodePrefix, encoding.EncodeUvarint(nil, uint64(nodeID)))
 }
 
-// MakeNamespaceMetadataKey returns the key for the namespace.
-func MakeNamespaceMetadataKey(namespace string) proto.Key {
-	return MakeKey(NamespaceMetadataPrefix, proto.Key(namespace))
+// MakeNameMetadataKey returns the key for the namespace.
+func MakeNameMetadataKey(parentID uint32, name string) proto.Key {
+	k := make([]byte, 0, len(NameMetadataPrefix)+10+len(name))
```
Comment (tamird) [inline keys/keys.go]: this `10` should really be a constant

File: client/namespace.go (hunk 3 above)
Comment (petermattis) [inline client/namespace.go]: This is part of the TODO at the top of the file to handle namespaces. I'll be changing in this in an upcoming PR to take a `namespace` argument.

File: client/namespace.go (hunk 1 above)
Comment (petermattis) [inline client/namespace.go]: Yeah, returning the struct was looking a bit ugly. Changed.

File: client/namespace.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+	schema = proto.TableSchemaFromDesc(desc)
+	return schema, nil
+}
+
+// RenameTable renames a table.
+func (db *DB) RenameTable(oldName, newName string) error {
+	// TODO(pmattis): Should we allow both the old and new name to exist
+	// simultaneously for a period of time? The thought is to allow an
+	// application to access the table via either name while the application is
+	// being upgraded. Alternately, instead of a rename table operation perhaps
```
Comment (MBerhault) [inline client/namespace.go]: s/Alternately/Alternatively/

File: client/namespace.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+)
+
+// TODO(pmattis):
+//
+// - This file contains the experimental Cockroach table-based interface. The
+//   API will eventually be moved into db.go, but is collected here during
+//   initial development. Much of the implementation will eventually wind up on
+//   the server using new table-based requests to perform operations.
+//
+// - Add support for namespaces. Currently namespace ID 0 is hardcoded.
```
Comment (tamird) [inline client/namespace.go]: is this the right file name?

File: client/namespace_test.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+	schema2, err := db.DescribeTable("USERS")
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !reflect.DeepEqual(schema, schema2) {
+		t.Errorf("expected %+v, but got %+v", schema, schema2)
+	}
+
+	// Creating the table again should fail as the table already exists.
+	if err := db.CreateTable(schema); err == nil {
```
Comment (tamird) [inline client/namespace_test.go]: could you use a stronger assertion here on the type of error you expect? this pattern is everywhere in our code but it's very very brittle

File: storage/store.go [hunk 8]
```diff
@@ -909,6 +909,15 @@ func (s *Store) BootstrapRange() error {
 	if err := engine.MVCCPutProto(batch, ms, key, now, nil, zoneConfig); err != nil {
 		return err
 	}
+
+	// We reserve the first 1000 descriptor IDs.
+	key = keys.DescIDGenerator
```
Comment (tamird) [inline storage/store.go]: would like to see a trivial test for this

**Comment by MBerhault** [comment]
LGTM

File: client/namespace.go (hunk 6 above)
Comment (petermattis) [inline client/namespace.go]: No. This really belongs in `table.go`. I'll move it there as soon as the `pmattis/client-table-impl` branch is merged.

File: client/namespace.go (hunk 3 above)
Comment (tamird) [inline client/namespace.go]: sgtm

File: client/namespace.go [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+		return err
+	} else if gr.Exists() {
+		return fmt.Errorf("table \"%s\" already exists", desc.Name)
+	}
+
+	ir, err := db.Inc(keys.DescIDGenerator, 1)
+	if err != nil {
+		return err
+	}
+	desc.ID = uint32(ir.ValueInt() - 1)
```
Comment (vivekmenezes) [inline client/namespace.go]: feels more natural to use the result rather than result - 1

Comment (petermattis) [inline client/namespace.go]: Perhaps, but our other `*IDGenerator` keys store the next ID to allocate, not the last ID allocated. I'd prefer consistency with the other ID generator keys.

File: client/namespace.go [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+			return err
+		}
+		desc.Name = strings.ToLower(newName)
+		if err := proto.ValidateTableDesc(desc); err != nil {
+			return err
+		}
+		newNameKey := keys.MakeNameMetadataKey(0, desc.Name)
+		b := &Batch{}
+		b.Put(descKey, &desc)
+		b.CPut(newNameKey, descKey, nil)
```
Comment (vivekmenezes) [inline client/namespace.go]: // If the new name already exists, the transaction fails

File: client/namespace.go (hunk 2 above)
Comment (vivekmenezes) [inline client/namespace.go]: I thought we were going to mark the table as deleted in its descriptor

Comment (petermattis) [inline client/namespace.go]: That's what the design doc says. See the TODO comment in `RenameTable` for my current thoughts on this.

File: client/namespace.go (hunk 10 above)
Comment (petermattis) [inline client/namespace.go]: Done.

File: client/namespace_test.go (hunk 7 above)
Comment (petermattis) [inline client/namespace_test.go]: I added a little helper method to check to see if the error matches a supplied regexp.

**Comment by vivekmenezes** [comment]
LGTM

File: storage/store.go (hunk 8 above)
Comment (petermattis) [inline storage/store.go]: Added a test into `TestCreateTable` that the newly created table has a non-reserved ID.

Comment (tamird) [inline storage/store.go]: :+1:

File: client/table.go [hunk 11]
```diff
@@ -196,6 +200,134 @@ func (m *model) encodeColumnKey(primaryKey []byte, column string) []byte {
...
+		return fmt.Errorf("table \"%s\" already exists", desc.Name)
+	}
+
+	ir, err := db.Inc(keys.DescIDGenerator, 1)
+	if err != nil {
+		return err
+	}
+	desc.ID = uint32(ir.ValueInt() - 1)
+
+	return db.Txn(func(txn *Txn) error {
```
Comment (spencerkimball) [inline client/table.go]: This is going to have an incomprehensible error message. You might consider adding a TODO here to be cognizant of error messages when this is ported to server-side code.

File: keys/keys.go (hunk 4 above)
Comment (petermattis) [inline keys/keys.go]: Added `maxVarintSize` constant.

File: client/namespace.go (hunk 5 above)
Comment (petermattis) [inline client/namespace.go]: Done.

**Comment by tamird** [comment]
LGTM. Would be nice to see the length of an `EncodeUvarint` come out as a constant.

File: client/namespace.go (hunk 6 above)
Comment (petermattis) [inline client/namespace.go]: This has now all been moved to `table.go` and the tests moved to `table_test.go`.

**Comment by spencerkimball** [comment]
LGTM

File: client/table.go (hunk 11 above)
Comment (petermattis) [inline client/table.go]: Ok. TODO added.

**Pull request merged**